src/utils/getTSType.js

```diff
@@ -0,0 +1,383 @@
+
+    type = {
+      ...type,
+      elements: params.map(param =>
+        getTSTypeWithResolvedTypes(param, typeParams),
+      ),
+      raw: printValue(path),
+    };
+  }
+
+  return type;
+}
+
+function getTSTypeWithRequirements(
+  path: NodePath,
+  typeParams: ?TypeParameters,
+): FlowTypeDescriptor {
+  const type = getTSTypeWithResolvedTypes(path, typeParams);
+  type.required = !path.parentPath.node.optional;
+  return type;
+}
+
+function handleTSTypeLiteral(
+  path: NodePath,
+  typeParams: ?TypeParameters,
+): FlowTypeDescriptor {
+  const type: FlowObjectSignatureType = {
+    name: 'signature',
+    type: 'object',
+    raw: printValue(path),
+    signature: { properties: [] },
+  };
+
+  path.get('members').each(param => {
+    if (
+      types.TSPropertySignature.check(param.node) ||
+      types.TSMethodSignature.check(param.node)
+    ) {
+      type.signature.properties.push({
+        key: getPropertyName(param),
+        value: getTSTypeWithRequirements(
+          param.get('typeAnnotation'),
+          typeParams,
+        ),
+      });
+    } else if (types.TSCallSignatureDeclaration.check(param.node)) {
+      type.signature.constructor = handleTSFunctionType(param, typeParams);
+    } else if (types.TSIndexSignature.check(param.node)) {
+      type.signature.properties.push({
+        key: getTSTypeWithResolvedTypes(
+          param
+            .get('parameters')
+            .get(0)
+            .get('typeAnnotation'),
+          typeParams,
+        ),
+        value: getTSTypeWithRequirements(
+          param.get('typeAnnotation'),
+          typeParams,
+        ),
+      });
+    }
+  });
+
+  return type;
+}
+
+function handleTSInterfaceDeclaration(path: NodePath): FlowElementsType {
+  // Interfaces are handled like references which would be documented separately,
+  // rather than inlined like type aliases.
+  return {
+    name: path.node.id.name,
+  };
+}
+
+function handleTSUnionType(
+  path: NodePath,
+  typeParams: ?TypeParameters,
+): FlowElementsType {
+  return {
+    name: 'union',
+    raw: printValue(path),
+    elements: path
+      .get('types')
+      .map(subType => getTSTypeWithResolvedTypes(subType, typeParams)),
+  };
+}
+
+function handleTSIntersectionType(
+  path: NodePath,
+  typeParams: ?TypeParameters,
+): FlowElementsType {
+  return {
+    name: 'intersection',
+    raw: printValue(path),
+    elements: path
+      .get('types')
+      .map(subType => getTSTypeWithResolvedTypes(subType, typeParams)),
+  };
+}
+
+function handleTSFunctionType(
+  path: NodePath,
+  typeParams: ?TypeParameters,
+): FlowFunctionSignatureType {
+  const type: FlowFunctionSignatureType = {
+    name: 'signature',
+    type: 'function',
+    raw: printValue(path),
+    signature: {
+      arguments: [],
+      return: getTSTypeWithResolvedTypes(
+        path.get('typeAnnotation'),
+        typeParams,
+      ),
+    },
+  };
+
+  path.get('parameters').each(param => {
```

Comment:
Should we treat the [`this` parameter](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters) different here and maybe assign it to it's own property `type.signature.this`? Just wonder because it is not a real parameter.
It probably does not make that much of a difference, as consumers of react-docgen could simply do that themself, but imho makes it a little bit easier to use.
This though would be the first difference to flow though if I'm correct :D